core/dataset.cpp

```diff
@@ -773,24 +780,24 @@ Dataset operator+(const Dataset &lhs, const DataConstProxy &rhs) {
   return res;
 }
 
-Dataset operator+(Dataset &&lhs, const Dataset &rhs) {
-  return apply(plus_equals, lhs, rhs);
+Dataset &&operator+(Dataset &&lhs, const Dataset &rhs) {
+  return apply(plus_equals, std::move(lhs), rhs);
 }
 
-Dataset operator+(Dataset &&lhs, const DatasetConstProxy &rhs) {
-  return apply(plus_equals, lhs, rhs);
+Dataset &&operator+(Dataset &&lhs, const DatasetConstProxy &rhs) {
+  return apply(plus_equals, std::move(lhs), rhs);
```

Comment:
I think implementing `+` using `+=` is not possible for `Dataset`. `+` may broadcast both arguments to compute the output, whereas `+=` cannot. Consider, e.g., two datasets, one depending on `Dim::X`, one on `Dim::Y`. The results is 2D data.